src/inventory/SlotSafeInventory.php

```diff
@@ -0,0 +1,45 @@
+<?php
+
+/*
+ *
+ *  ____            _        _   __  __ _                  __  __ ____
+ * |  _ \ ___   ___| | _____| |_|  \/  (_)_ __   ___      |  \/  |  _ \
+ * | |_) / _ \ / __| |/ / _ \ __| |\/| | | '_ \ / _ \_____| |\/| | |_) |
+ * |  __/ (_) | (__|   <  __/ |_| |  | | | | | |  __/_____| |  | |  __/
+ * |_|   \___/ \___|_|\_\___|\__|_|  |_|_|_| |_|\___|     |_|  |_|_|
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * @author PocketMine Team
+ * @link http://www.pocketmine.net/
+ *
+ *
+ */
+
+declare(strict_types=1);
+
+namespace pocketmine\inventory;
+
+use pocketmine\inventory\transaction\validator\InventoryTransactionValidator;
+use pocketmine\utils\ObjectSet;
+
+/**
+ * A "slot safe inventory" is an inventory that has a set of rules that determine whether an item can be placed in a
+ * particular slot. This ensures that the inventory's internal state is always valid and consistent.
+ */
+interface SlotSafeInventory{
+	/**
+	 * Return a set of validators that will be used to determine whether an item can be placed in a particular slot.
+	 * All validators need to return true for the transaction to be allowed.
+	 * If one of the validators returns false, the transaction will be cancelled.
+	 *
+	 * There is no guarantee that the validators will be called in any particular order.
+	 * All validators need to be stateless and not depend on the order in which they are called.
+	 *
+	 * @phpstan-return ObjectSet<InventoryTransactionValidator>
+	 */
+	public static function getSlotValidators() : ObjectSet;
```

Comment:
why static? this seems unnecessarily limiting

Reply:
So plugins can add any validator to the set of an inventory. 
Example: you want your custom item to be added on the head slot, you call the `getSlotValidators` on the load or enable of the plugins so it will be added all the time.

Reply:
I can't see a case where we'd need to do this for a non custom inventory. For a custom one we'd just override the constructor.

This API design seems weird and it creates unnecessary limitations in other areas.

Reply:
> I can't see a case where we'd need to do this for a non custom inventory. For a custom one we'd just override the constructor.

If any case, you want to put a red carpet on your head because you created a custom logic for this block. With the new validator you can't except if you can a validator to support then.

Reply:
Static may not be the best solution, but we need to allow plugins adding validators.

Reply:
I agree, but I don't think static is the way to do it.

Typically I would think EntitySpawnEvent or ChunkLoadEvent would be sufficient to setup validators for vanilla things.

Reply:
Either that or just apply validators as-needed during InventoryOpenEvent... that would probably make more sense